### ISSUE TITLE
box_collection lint shouldn't trigger when used in `extern "C"` functions

### ISSUE THREAD

**Issue by morrisonlevi** [rust-lang/rust-clippy#8587]
### Summary

Clippy warns about `Box<Vec<T>>` as being a useless heap allocation, but when they are used in `extern "C"` functions the Box is necessary as Vec's layout isn't `repr(C)`.

### Lint Name

box_collection

### Reproducer

I tried this code:

```rust
#[no_mangle]
pub extern "C" fn vec_tag_new() -> Box<Vec<u32>> {
    Box::new(Vec::new())
}
```

I saw this happen:

```
error: you seem to be trying to use `Box<Vec<..>>`. Consider using just `Vec<..>`
   --> src/lib.rs:283:36
    |
283 | pub extern "C" fn vec_tag_new() -> Box<Vec<u32>> {
    |                                    ^^^^^^^^^^^^^
    |
    = note: `-D clippy::box-collection` implied by `-D warnings`
    = help: `Vec<..>` is already on the heap, `Box<Vec<..>>` makes an extra allocation
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#box_collection
```

I expected to see this happen:

No lint.

### Version

```text
rustc 1.59.0 (9d1b2106e 2022-02-23)
binary: rustc
commit-hash: 9d1b2106e23b1abd32fce1f17267604a5102f57a
commit-date: 2022-02-23
host: x86_64-apple-darwin
release: 1.59.0
LLVM version: 13.0.0
```

**Comment by J-ZhengLi** [comment]
Hmmmm, it shouldn't be triggering this lint and I just test clippy with your code on [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d734799423c2bba997e983ba43da24da) but got no error? Can you try it again and provide us with more context?
